ext/ImplicitDifferentiationChainRulesExt.jl

```diff
@@ -24,59 +26,54 @@ function ChainRulesCore.rrule(
     rc::RuleConfig,
     implicit::ImplicitFunction,
     x::AbstractArray{R},
-    ::Val{return_byproduct};
+    ::ReturnByproduct;
     kwargs...,
-) where {R,return_byproduct}
+) where {R}
     @unpack conditions, linear_solver = implicit
 
-    y, z = implicit(x, Val(true); kwargs...)
+    y, z = implicit(x, ReturnByproduct(); kwargs...)
     n, m = length(x), length(y)
 
     backend = ReverseRuleConfigBackend(rc)
     pbA = pullback_function(backend, _y -> conditions(x, _y, z; kwargs...), y)
     pbB = pullback_function(backend, _x -> conditions(_x, y, z; kwargs...), x)
     pbmA = PullbackMul!(pbA, size(y))
     pbmB = PullbackMul!(pbB, size(y))
+
     Aᵀ_op = LinearOperator(R, m, m, false, false, pbmA)
     Bᵀ_op = LinearOperator(R, n, m, false, false, pbmB)
-    implicit_pullback = ImplicitPullback(
-        Aᵀ_op, Bᵀ_op, linear_solver, x, Val(return_byproduct)
-    )
+    Aᵀ_op_presolved = presolve(linear_solver, Aᵀ_op, y)
+
+    implicit_pullback = ImplicitPullback(Aᵀ_op_presolved, Bᵀ_op, linear_solver, x)
+
+    return (y, z), implicit_pullback
+end
 
-    if return_byproduct
-        return (y, z), implicit_pullback
-    else
-        return y, implicit_pullback
-    end
+function ChainRulesCore.rrule(
+    rc::RuleConfig, implicit::ImplicitFunction, x::AbstractArray{R}; kwargs...
+) where {R}
+    (y, z), implicit_pullback = rrule(rc, implicit, x, ReturnByproduct(); kwargs...)
+    implicit_pullback_no_byproduct(dy) = implicit_pullback((dy, nothing))
+    return y, implicit_pullback_no_byproduct
 end
 
-struct ImplicitPullback{return_byproduct,A,B,L,X}
+struct ImplicitPullback{A,B,L,X}
     Aᵀ_op::A
     Bᵀ_op::B
     linear_solver::L
     x::X
-    _v::Val{return_byproduct}
-end
-
-function (implicit_pullback_nobyproduct::ImplicitPullback{false})(dy)
-    @unpack Aᵀ_op, Bᵀ_op, linear_solver, x = implicit_pullback_nobyproduct
-    implicit_pullback_byproduct = ImplicitPullback(
-        Aᵀ_op, Bᵀ_op, linear_solver, x, Val(true)
-    )
-    return implicit_pullback_byproduct((dy, nothing))
 end
 
-function (implicit_pullback_byproduct::ImplicitPullback{true})((dy, _))
-    @unpack Aᵀ_op, Bᵀ_op, linear_solver, x = implicit_pullback_byproduct
+function (implicit_pullback::ImplicitPullback)((dy, dz))
+    @unpack Aᵀ_op, Bᵀ_op, linear_solver, x = implicit_pullback
     R = eltype(x)
-
-    dy_vec = convert(Vector{R}, vec(unthunk(dy)))
-    dF_vec, stats = linear_solver(Aᵀ_op, dy_vec)
-    check_solution(linear_solver, stats)
-    dx_vec = Bᵀ_op * dF_vec
-    dx_vec .*= -1
+    dy_vec = convert(AbstractVector{R}, vec(unthunk(dy)))
+    dF_vec = solve(linear_solver, Aᵀ_op, dy_vec)
+    dx_vec = vec(similar(x))
+    mul!(dx_vec, Bᵀ_op, dF_vec)
+    lmul!(-one(R), dx_vec)
     dx = reshape(dx_vec, size(x))
-    return (NoTangent(), dx, NoTangent())
+    return (NoTangent(), dx)
```

Comment:
The old implementation was correct for the `(f::ImplicitFunction)(x, ReturnByProduct())` since `rrule` needs to return 3 co-tangents: one for the function (`NoTangent()`), one for `x` (`dx`) and one for `ReturnByProduct()` (`NoTangent()`). The new implementation is correct for `(f::ImplicitFunction)(x)`. Since the `rrule` for the no-byproduct call is defined in terms of the `rrule` for the byproduct call:
```julia
function ChainRulesCore.rrule(
    rc::RuleConfig, implicit::ImplicitFunction, x::AbstractArray{R}; kwargs...
) where {R}
    (y, z), implicit_pullback = rrule(rc, implicit, x, ReturnByproduct(); kwargs...)
    implicit_pullback_no_byproduct(dy) = implicit_pullback((dy, nothing))
    return y, implicit_pullback_no_byproduct
end
```
I think we should make `implicit_pullback` return 3 outputs by default and just replace this line
```julia
implicit_pullback_no_byproduct(dy) = implicit_pullback((dy, nothing))
```
with
```julia
implicit_pullback_no_byproduct(dy) = Base.front(implicit_pullback((dy, nothing)))
```

Reply:
The fact that this bug wasn't caught in the tests is concerning, I think we need to test differentiating `f(g(h(x)))` where `g` is an implicit function to ensure all tangents and co-tangents are used correctly.

Reply:
Oooh right that's where the three tangents come from. I'll put them back in.
The systematic tests don't include byproducts yet, but now that I've found a simple enough example I'll put that in

Reply:
Putting waaaaay more tests in the next commit